pkg/apis/configuration/validation/validation.go

```diff
@@ -105,6 +115,82 @@ func validateUpstreamLBMethod(lBMethod string, fieldPath *field.Path, isPlus boo
 	return allErrs
 }
 
+func validateUpstreamHealthCheck(hc *v1alpha1.HealthCheck, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if hc == nil || !hc.Enable {
+		return allErrs
+	}
+
+	allErrs = append(allErrs, validatePath(hc.Path, fieldPath.Child("path"))...)
+	allErrs = append(allErrs, validateTime(hc.Interval, fieldPath.Child("interval"))...)
+	allErrs = append(allErrs, validateTime(hc.Jitter, fieldPath.Child("jitter"))...)
+	allErrs = append(allErrs, validatePositiveInt(hc.Fails, fieldPath.Child("fails"))...)
+	allErrs = append(allErrs, validatePositiveInt(hc.Passes, fieldPath.Child("passes"))...)
+	allErrs = append(allErrs, validateTime(hc.ConnectTimeout, fieldPath.Child("connect-timeout"))...)
+	allErrs = append(allErrs, validateTime(hc.ReadTimeout, fieldPath.Child("read-timeout"))...)
+	allErrs = append(allErrs, validateTime(hc.SendTimeout, fieldPath.Child("send-timeout"))...)
+	allErrs = append(allErrs, validateStatusMatch(hc.StatusMatch, fieldPath.Child("statusMatch"))...)
+
+	for i, header := range hc.Headers {
+		idxPath := fieldPath.Child("headers").Index(i)
+		allErrs = append(allErrs, validateHeader(header, idxPath)...)
+	}
+
+	if hc.Port > 0 {
+		for _, msg := range validation.IsValidPortNum(int(hc.Port)) {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("port"), hc.Port, msg))
+		}
+	}
+
+	return allErrs
+}
+
+const statusMatchFmt = "(\\d*|\\s|\\-|\\!\\s)*"
+const statusMatchFmtErrMsg string = "a valid statusMatch must consist of digits, spaces, `! ` or `-`"
+
+var statusMatchRegexp = regexp.MustCompile("^" + statusMatchFmt + "$")
+
+func validateStatusMatch(s string, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !statusMatchRegexp.MatchString(s) {
+		msg := validation.RegexError(statusMatchFmtErrMsg, statusMatchFmt, "200", "! 500", "200-399")
+		return append(allErrs, field.Invalid(fieldPath, s, msg))
+	}
+	return allErrs
+}
+
+func validateHeader(h v1alpha1.Header, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if h.Name == "" {
+		allErrs = append(allErrs, field.Required(fieldPath.Child("name"), ""))
+	}
+
+	for _, msg := range validation.IsHTTPHeaderName(h.Name) {
+		allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), h.Name, msg))
+	}
+
+	for _, msg := range isValidHeaderValue(h.Value) {
+		allErrs = append(allErrs, field.Invalid(fieldPath.Child("value"), h.Value, msg))
+	}
+
+	return allErrs
+}
+
+const headerValueFmt = "[^{};'\"]*"
```

Comment:
I think here we need to have a validation similar to machValue - a valid value must have all '"' (double quotes) escaped and must not end with an unescaped '\' (backslash). And additionally, we can disallow using variables by banning `$` sign. Note, for health check most of NGINX built in variables will be empty, because they are typically related to request/response.

Reply:
Sounds good